File: 242.valid-anagram.py
#
# @lc app=leetcode id=242 lang=python
#
# [242] Valid Anagram
#

# @lc code=start

class Solution(object):
    def isAnagram(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        
        # 1st Solution
        # return sorted(s) == sorted(t)

        # 2nd Solution
        # Store the count of each character in dictionary
        letters = {}
        for char in s:
            if char in letters: letters[char] += 1
            else: letters[char] = 1
        
        # If the character is in t, decrement the count
        for char in t:
            if char in letters: letters[char] -= 1
            else: return False
        
        # Check that the count of each character is 0
        for value in letters.values():
            if value != 0: return False
        
        return True

        
# @lc code=end

File: 242.valid-anagram.cpp
/*
 * @lc app=leetcode id=242 lang=cpp
 *
 * [242] Valid Anagram
 */

// @lc code=start
#include <string>;
#include <unordered_map>;
using namespace std;

class Solution {
public:
    bool isAnagram(string s, string t) {
        if (s.size() != t.size()) return false;
        
        unordered_map<char, int> frequency;
        for (char c : s) frequency[c]++;
        for (char c : t){
            if (frequency.find(c) == frequency.end()) return false;
            frequency[c]--;
            if (frequency[c] == 0) frequency.erase(c);
        }
        return frequency.empty();
    }
};
// @lc code=end

